src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/AttributePresenceFilterNode.cs

```diff
@@ -0,0 +1,295 @@
+                        continue;
+                    }
+
+                    // Determine if there is space in a bucket to add a new entry
+                    if (isEmptyEntryInBucket(bucketAIndex))
+                    {
+                        fillEmptyEntryInBucket(bucketAIndex, fingerprint);
+                        continue;
+                    }
+
+                    if (isEmptyEntryInBucket(bucketBIndex))
+                    {
+                        fillEmptyEntryInBucket(bucketBIndex, fingerprint);
+                        continue;
+                    }
+
+                    bool success = false;
+                    int MaxNumKicks = 256;
+                    // Note, that bucketAIndex itself was chosen randomly above.
+                    for (int n = 0; !success && n < MaxNumKicks; n++)
+                    {
+                        // Randomly swap an entry in bucket bucketAIndex with fingerprint
+                        uint entryIndexInBucket = XorShift128(state) & 0x7;
+                        ushort temp = fingerprint;
+                        fingerprint = pTable[(bucketAIndex * 8) + entryIndexInBucket];
+                        pTable[(bucketAIndex * 8) + entryIndexInBucket] = temp;
+
+                        // Find other bucket
+                        fingerprintHash = (uint)fingerprint;
+                        bucketAIndex = bucketAIndex ^ (fingerprintHash % bucketCount);
+                        if (isEmptyEntryInBucket(bucketAIndex))
+                        {
+                            fillEmptyEntryInBucket(bucketAIndex, fingerprint);
+                            continue;
+                        }
+                    }
+
+                    if (success)
+                    {
+                        continue;
+                    }
+
+                    tryAgainWithBiggerTable = true;
+                }
+
+                if (tryAgainWithBiggerTable)
+                {
+                    // bucket entry kicking path requires bucket counts to be power of two in size due to use of xor to retrieve second hash
+                    bucketCount *= 2;
+                }
+            }
+            while(tryAgainWithBiggerTable && ((countOfRetries++) < 2));
+
+            byte[] result;
+            if (tryAgainWithBiggerTable)
+            {
+                result = Array.Empty<byte>();
+            }
+            else
+            {
+                result = new byte[pTable.Length * 2];
+                for (int i = 0; i < pTable.Length; i++)
+                {
+                    result[i * 2] = (byte)(pTable[i] % 256);
+                    result[i * 2 + 1] = (byte)(pTable[i] >> 8);
+                }
+            }
+
+            ObjectDataBuilder builder = new ObjectDataBuilder(factory, relocsOnly);
+            builder.RequireInitialAlignment(16);
+            builder.AddSymbol(this);
+            builder.EmitBytes(result);
+            computedData = builder.ToObjectData();
```

Comment:
We don't need to cache `ObjectData`. You have an early out for `relocsOnly` so this will only be called once.

Reply:
This is not true, the method is called twice with `relocsOnly == false` with the following stacks:

```
>	ILCompiler.ReadyToRun.dll!ILCompiler.DependencyAnalysis.ReadyToRun.AttributePresenceFilterNode.GetData(ILCompiler.DependencyAnalysis.NodeFactory factory, bool relocsOnly) Line 123	C#
 	ILCompiler.ReadyToRun.dll!ILCompiler.DependencyAnalysis.ReadyToRun.HeaderNode.GetData(ILCompiler.DependencyAnalysis.NodeFactory factory, bool relocsOnly) Line 139	C#
 	ILCompiler.ReadyToRun.dll!ILCompiler.DependencyAnalysis.ReadyToRunObjectWriter.EmitPortableExecutable() Line 93	C#
 	ILCompiler.ReadyToRun.dll!ILCompiler.DependencyAnalysis.ReadyToRunObjectWriter.EmitObject(System.Reflection.PortableExecutable.PEReader inputPeReader, string objectFilePath, System.Collections.Generic.IEnumerable<ILCompiler.DependencyAnalysisFramework.DependencyNode> nodes, ILCompiler.DependencyAnalysis.ReadyToRunCodegenNodeFactory factory) Line 201	C#
 	ILCompiler.ReadyToRun.dll!ILCompiler.ReadyToRunCodegenCompilation.Compile(string outputFile) Line 224	C#
 	crossgen2.dll!ILCompiler.Program.Run(string[] args) Line 403	C#
 	crossgen2.dll!ILCompiler.Program.Main(string[] args) Line 479	C#

```

and then once again

```
>	ILCompiler.ReadyToRun.dll!ILCompiler.DependencyAnalysis.ReadyToRun.AttributePresenceFilterNode.GetData(ILCompiler.DependencyAnalysis.NodeFactory factory, bool relocsOnly) Line 123	C#
 	ILCompiler.ReadyToRun.dll!ILCompiler.DependencyAnalysis.ReadyToRun.HeaderNode.GetData(ILCompiler.DependencyAnalysis.NodeFactory factory, bool relocsOnly) Line 139	C#
 	ILCompiler.ReadyToRun.dll!ILCompiler.DependencyAnalysis.ReadyToRun.CopiedCorHeaderNode.GetData(ILCompiler.DependencyAnalysis.NodeFactory factory, bool relocsOnly) Line 127	C#
 	ILCompiler.ReadyToRun.dll!ILCompiler.DependencyAnalysis.ReadyToRunObjectWriter.EmitPortableExecutable() Line 93	C#
 	ILCompiler.ReadyToRun.dll!ILCompiler.DependencyAnalysis.ReadyToRunObjectWriter.EmitObject(System.Reflection.PortableExecutable.PEReader inputPeReader, string objectFilePath, System.Collections.Generic.IEnumerable<ILCompiler.DependencyAnalysisFramework.DependencyNode> nodes, ILCompiler.DependencyAnalysis.ReadyToRunCodegenNodeFactory factory) Line 201	C#
 	ILCompiler.ReadyToRun.dll!ILCompiler.ReadyToRunCodegenCompilation.Compile(string outputFile) Line 224	C#
 	crossgen2.dll!ILCompiler.Program.Run(string[] args) Line 403	C#
 	crossgen2.dll!ILCompiler.Program.Main(string[] args) Line 479	C#
```

Reply:
We shouldn't cache this. Nothing else in our system does such caching. The more correct question is why do we call this twice. We shouldn't do that either. Please file an issue against @nattress

Reply:
If its just a size computation reason for why this is happening, I suppose the cache is ok.

Reply:
The R2R header encodes node RVAs and lengths so we call GetData to emit the node into the image and also to get its size for the header. We should be able to add a relocation-like feature to the object writer which avoids the double copy and fixes up node lengths after laying them out.

Reply:
Issue #27166 is filed to capture the work to eliminate a `GetData()` call.